src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,438 @@
+    frameBuffer_.attachRenderbuffer(
+        Mn::GL::Framebuffer::BufferAttachment::Depth, optionalDepthBuffer_);
+  }
+}
+
+void CubeMap::recreateTexture() {
+  Mn::Vector2i size{imageSize_, imageSize_};
+
+  // color texture
+  if (flags_ & Flag::ColorTexture) {
+    auto& colorTexture = textures_[TextureType::Color];
+    colorTexture = std::make_unique<Mn::GL::CubeMapTexture>();
+    (*colorTexture)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setMinificationFilter(Mn::GL::SamplerFilter::Linear,
+                               Mn::GL::SamplerMipmap::Linear)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Linear);
+
+    if (flags_ & Flag::BuildMipmap) {
+      // RGBA8 is for the LDR. Use RGBA16F for the HDR (TODO)
+      (*colorTexture)
+          .setStorage(Mn::Math::log2(imageSize_) + 1,
+                      Mn::GL::TextureFormat::RGBA8, size);
+    } else {
+      (*colorTexture).setStorage(1, Mn::GL::TextureFormat::RGBA8, size);
+    }
+  }
+
+  // depth texture
+  if (flags_ & Flag::DepthTexture) {
+    auto& depthTexture = textures_[TextureType::Depth];
+    depthTexture = std::make_unique<Mn::GL::CubeMapTexture>();
+    (*depthTexture)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setStorage(1, Mn::GL::TextureFormat::DepthComponent32F, size);
+  }
+}
+
+void CubeMap::recreateFramebuffer() {
+  Mn::Vector2i viewportSize{imageSize_, imageSize_};
+  frameBuffer_ = Mn::GL::Framebuffer{{{}, viewportSize}};
+  // optional depth buffer is 24-bit integer pixel, which is different from the
+  // depth texture (32-bit float)
+  optionalDepthBuffer_.setStorage(Mn::GL::RenderbufferFormat::DepthComponent24,
+                                  viewportSize);
+}
+
+void CubeMap::prepareToDraw(unsigned int cubeSideIndex) {
+  mapForDraw(cubeSideIndex);
+
+  // sorry, unlike color buffers, for depth buffer you have to reattach it every
+  // time
+  // however, if NOT using depth texture, we do not need to attach the depth
+  // buffer again and again
+  if (flags_ & Flag::DepthTexture) {
+    Magnum::GL::CubeMapCoordinate cubeMapCoord =
+        convertFaceIndexToCubeMapCoordinate(cubeSideIndex);
+    frameBuffer_.attachCubeMapTexture(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        *textures_[TextureType::Depth], cubeMapCoord, 0);
+  }
+
+  frameBuffer_.clearDepth(1.0f).clearColor(0,                // color attachment
+                                           Mn::Vector4ui{0}  // clear color
+  );
+
+  CORRADE_INTERNAL_ASSERT(
+      frameBuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+}
+
+void CubeMap::mapForDraw(unsigned int colorAttachmentIndex) {
+  frameBuffer_.mapForDraw({
+      {Mn::Shaders::Generic3D::ColorOutput,
+       colorAttachment[colorAttachmentIndex]},
```

Comment:
Why have an array for this and not directly use

```suggestion
       Mn::GL::Framebuffer::ColorAttachment{colorAttachmentIndex}},
```

? The array indirection only complicates that in my opinion.

Hm and given that you need to re-attach the depth buffer every time, I'm not sure about the efficiency of this suggested approach either...

Reply:
It still helps in the color sensor case. The performance got visibly improved. See the following test results:
Before:
![before-change-fps](https://user-images.githubusercontent.com/931093/105771947-45ce4f80-5f16-11eb-8d4f-bff8d5dd1369.png)

After:
![after-change-fps](https://user-images.githubusercontent.com/931093/105771958-4c5cc700-5f16-11eb-8c63-90f313147bbe.png)

Reply:
> The array indirection only complicates that in my opinion.

Yes, I see your point. I defined them in the very beginning of this file and by doing so, later if we would like to define the another color attachment for the objectId,  it is less error prone. (I would remember that the first 6 color attachments are already taken.)

Reply:
Oh nice, the frame profiler is already useful :heart: 

If you have time, could you try how the CPU overhead would be with six different framebuffer objects (my original suggestion)?

Reply:
Yes, I did the tests:
1 FBO + 6 colorAttachements VS 6FBOs

In "1FBO + 6": It needs to attach depth buffer every frame. 
(Note: I also have fbo.bind(...) in every frame but since the fbo never changes, I assume Magnum will NOT actually bind it again and again.)

In 6FBOs: It does NOT need to attach any color or depth buffer, but need to rebind FBOs each time.

The following pic is for "1FBO+6":
![1_6](https://user-images.githubusercontent.com/931093/108161083-2b177280-709f-11eb-80dc-10c7e13d725a.jpeg)

The following pic is for "6FBOs":
![6fbos](https://user-images.githubusercontent.com/931093/108161128-42eef680-709f-11eb-834e-d16ec73c7259.jpeg)

In terms of CPU time: 1FBO+6 < 6FBOs
In terms of GPU time: 1FBO+6 > 6FBOs
Overall time: 1FBO+6 > 6FBOs (winner)
This AB test has been repeated multiple times, and I used my cellphone to capture the results (screenshot app will significantly affect the measurement.). The results are trustable.

@mosra: Can you explain why 6FBOs beats 1FBO+6 on GPU?

Reply:
The PR is updated to use the 6FBOs based on the result.